### PR TITLE
[#113] 사용자는 영화의 시간 스케줄을 확인한다

### PR DESCRIPTION
**Pull request by yanJuicy**
## 개요
- 상영시간 조회 기능

## 추가기능
- 영화 ID, 영화관 ID, 날짜를 요청하여 상영관, 총 좌석 수, 상영 시간 리스트 확인

## 변경사항(Optional)
- Message enum 에러 메시지 추가
    - `INVALID_MOVIE_EXP_MSG("존재하지 않는 영화 ID 입니다.")`
    - `INVALID_THEATER_ROOM_EXP_MSG("존재하지 않는 상영관 ID 입니다.")`

## 사용방법(Optional)
- 로그인을 하지 않아도 조회 기능을 사용할 수 있다.

## 기타
<h3 id="_영화_영화관_상영_날짜를_요청하여_상영_시간_리스트_가져오기" style="box-sizing: border-box; margin: 1em 0px 0.5em; padding: 0px; font-family: &quot;Open Sans&quot;, &quot;DejaVu Sans&quot;, sans-serif; font-weight: 300; font-style: normal; color: rgb(204, 204, 204); text-rendering: optimizelegibility; line-height: 1.2; font-size: 1.375em; word-spacing: -0.05em; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">영화, 영화관, 상영 날짜를 요청하여 상영 시간 리스트 가져오기</h3><div class="listingblock has-source-line data-line-/Users/hanju/IdeaProjects/BE-02-MarBox/build/generated-snippets/schedule-get-timeList-by-movie-and-theater-and-date/http-request.adoc-2" style="box-sizing: border-box; margin: 0px 0px 1.25em; padding: 0px; color: rgb(187, 187, 187); font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="title" style="box-sizing: border-box; margin: 0px 0px 0.25em; padding: 0px; line-height: 1.45; color: rgb(122, 37, 24); font-weight: 400; text-rendering: optimizelegibility; text-align: left; font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 1rem; font-style: italic; filter: invert(90%) hue-rotate(180deg);">Request</div><div class="content" style="box-sizing: border-box; margin: 0px; padding: 0px; position: relative;"><pre class="CodeRay highlight nowrap" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: 0.8125em; white-space: pre; margin: 0px; padding: 1em; color: rgba(0, 0, 0, 0.9); line-height: 1.45; text-rendering: optimizespeed; overflow-wrap: normal; background: rgb(247, 247, 248); filter: invert(90%) hue-rotate(180deg); border-radius: 4px; overflow-x: auto;"><code data-lang="http" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: inherit; font-weight: 400; color: inherit; line-height: inherit; display: block;">GET /schedules?movieId=1&amp;theaterId=1&amp;date=2022-07-01 HTTP/1.1
Host: localhost:8080</code></pre></div></div><div class="listingblock has-source-line data-line-/Users/hanju/IdeaProjects/BE-02-MarBox/build/generated-snippets/schedule-get-timeList-by-movie-and-theater-and-date/http-response.adoc-2" style="box-sizing: border-box; margin: 0px 0px 1.25em; padding: 0px; color: rgb(187, 187, 187); font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="title" style="box-sizing: border-box; margin: 0px 0px 0.25em; padding: 0px; line-height: 1.45; color: rgb(122, 37, 24); font-weight: 400; text-rendering: optimizelegibility; text-align: left; font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 1rem; font-style: italic; filter: invert(90%) hue-rotate(180deg);">Response</div><div class="content" style="box-sizing: border-box; margin: 0px; padding: 0px; position: relative;"><pre class="CodeRay highlight nowrap" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: 0.8125em; white-space: pre; margin: 0px; padding: 1em; color: rgba(0, 0, 0, 0.9); line-height: 1.45; text-rendering: optimizespeed; overflow-wrap: normal; background: rgb(247, 247, 248); filter: invert(90%) hue-rotate(180deg); border-radius: 4px; overflow-x: auto;"><code data-lang="http" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: inherit; font-weight: 400; color: inherit; line-height: inherit; display: block;">HTTP/1.1 200 OK
Content-Type: application/json
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
X-Frame-Options: DENY
Content-Length: 258

{"movieList":[],"theaterList":[],"dateList":[],"timeList":[{"theaterRoomName":"상영관1","totalSeatCount":75,"startTimeList":["09:30:00","14:25:00","16:45:00","19:30:00"]},{"theaterRoomName":"상영관2","totalSeatCount":100,"startTimeList":["20:40:00"]}]}</code></pre></div></div>

Path | Type | Description
-- | -- | --
timeList[] | Array | 상영 시간 정보
timeList[].theaterRoomName | String | 상영관 이름
timeList[].totalSeatCount | Number | 상영관 총 좌석 수
timeList[].startTimeList[] | Array | 상영관의 영화 시작 시간 배열
movieList[] | Array | 빈 배열
theaterList[] | Array | 빈 배열
dateList[] | Array | 빈 배열

### CODE REVIEW

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java [hunk 1]
```diff
@@ -118,6 +119,31 @@ public ResponseFindSchedule findMovieListByTheaterIdAndDate(Long theaterId, Loca
...
+	@Transactional(readOnly = true)
+	public ResponseFindSchedule findTimeScheduleList(Long movieId, Long theaterId, LocalDate date) {
+		movieRepository.findById(movieId)
+			.orElseThrow(() -> new IllegalArgumentException(Message.INVALID_MOVIE_EXP_MSG.getMessage()));
+		theaterRepository.findById(theaterId)
+			.orElseThrow(() -> new EntityNotFoundException(Message.INVALID_THEATER_EXP_MSG.getMessage()));
+		if (LocalDate.now().plusDays(CURRENT_SCHEDULE_PERIOD).isBefore(date) || LocalDate.now().isAfter(date)) {
+			throw new DateTimeException(Message.INVALID_DATE_EXP_MSG.getMessage());
+		}
+
```
Comment (wisehero) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java:131]: 이 부분  다른 메서드에도 중복인 것 같은데 따로 메서드로 빼면 어떨까 싶습니다.

Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java:131]: 넵!  파라미터로 전달되는 값을 확인하는 코드가 계속 중복되서 따로 빼는게 좋을 것 같습니다

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java [hunk 2]
```diff
@@ -30,4 +34,22 @@ public List<LocalDate> convertFromScheduleListToDateList(List<Schedule> schedule
 			.toList();
 	}
 
+	public List<ResponseFindTime> convertFromScheduleInTheaterToResponseFindTimeList(
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java:37]: 복잡하네요 .. 고생하셨습니다 !
로직이 좀 복잡해보여서 간단한 주석이 있다면 더 좋을것 같네요 ㅎ

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java [hunk 3]
```diff
@@ -118,6 +119,31 @@ public ResponseFindSchedule findMovieListByTheaterIdAndDate(Long theaterId, Loca
...
+		List<Schedule> schedulesOfAllTheaters = scheduleRepository.findSchedulesByMovieIdAndDate(movieId, date);
+		Set<TheaterRoom> theaterRoomsOfRequestTheater = theaterRoomRepository.findAllByTheaterId(theaterId);
+
+		List<Schedule> schedulesInRequestTheater = schedulesOfAllTheaters.stream()
+			.filter(schedule -> theaterRoomsOfRequestTheater.contains(schedule.getTheaterRoom()))
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java:136]: findSchedulesByMovieIdAndDate함수에서 TheaterRoom은 fetch join이 적용안된걸로 보이는데
N + 1 문제가 발생하지는 않나요 ??

File: src/test/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleServiceTest.java [hunk 4]
```diff
@@ -247,6 +246,42 @@ void testFindMovieListByTheaterIdAndDate_Fail_Invalid_Date() {
...
+		createAndSaveSchedule(theaterRoom2, movie,
+			LocalDateTime.of(LocalDate.now(), LocalTime.of(9, 0)),
+			LocalDateTime.of(LocalDate.now(), LocalTime.of(11, 0)));
+		createAndSaveSchedule(theaterRoom2, movie2,
+			LocalDateTime.of(LocalDate.now(), LocalTime.of(12, 30)),
+			LocalDateTime.of(LocalDate.now(), LocalTime.of(14, 30)));
+		createAndSaveSchedule(theaterRoom2, movie,
+			LocalDateTime.of(LocalDate.now().plusDays(1), LocalTime.of(12, 30)),
+			LocalDateTime.of(LocalDate.now().plusDays(1), LocalTime.of(14, 30)));
+
```
Comment (Pawer0223) [inline src/test/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleServiceTest.java:275]: 여기서 flush, clear를 수행해보면 아래 service에서 theater_room에 접근할때마다 select가 나갈것 같은데.. 맞나요 ?

**Review by Pawer0223 — COMMENTED** [review]
고생하셨습니다 !

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java (hunk 2 above)
Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java:37]: 좀 더 보기 좋게 재구조 해보겠습니다!

File: src/test/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleServiceTest.java (hunk 4 above)
Comment (yanJuicy) [reply src/test/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleServiceTest.java:275]: 확인해보고 바꾸겠습니다!

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java [hunk 5]
```diff
@@ -34,22 +35,30 @@ public List<LocalDate> convertFromScheduleListToDateList(List<Schedule> schedule
...
+				getStartTimeListOfSchedulesInTheaterRoom(schedules, theaterRoom);
+
+			responseFindTimeSchedules.add(new ResponseFindTime(theaterRoom.getName(),
+				theaterRoom.getTotalAmount(), startTimeListInTheaterRoom));
 		}
 
 		return responseFindTimeSchedules;
 	}
 
+	private List<LocalTime> getStartTimeListOfSchedulesInTheaterRoom(List<Schedule> schedules,
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java:56]: 개인적으로 이전보다 convertFromScheduleListToResponseFindTimeList에서 무엇을 하는지 파악하기 좋아진것 같습니다 👍

File: src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java [hunk 6]
```diff
@@ -23,6 +23,7 @@ public interface ScheduleRepository extends JpaRepository<Schedule, Long> {
 	List<Schedule> findScheduleByDate(LocalDate date);
 
 	@Query("SELECT s FROM Schedule s "
+		+ "JOIN FETCH s.theaterRoom "
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java:26]: 👍 직접 코드를 돌려보지 않고 추측했던건데, 실제로 N+1문제가 발생했던가요 ?

**Review by Pawer0223 — APPROVED** [review]
LGTM

File: src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java (hunk 6 above)
Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java:26]: @Pawer0223  
네 fetch join 없었을 때 보니까 theaterRoom을 확인하는 select 문이 발생하더라고요!

**Review by wisehero — APPROVED** [review]
LGTM

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java [hunk 7]
```diff
@@ -53,7 +53,9 @@ public ResponseEntity<ResponseFindSchedule> getSchedulesByRequestParam(
 
 		ResponseFindSchedule responseFindSchedule;
 
-		if (movieId == null && theaterId != null && date == null) {
+		if (movieId != null && theaterId != null && date != null) {
```
Comment (ksy90101) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:56]: 으음... if문이 너무 많네요?
이런경우에는 항상 이야기 하는 전략패턴을 충분히 사용 가능할거 같은데 한번 해보면 어떨까요?
아울러 조건문을 조금 더 stream을 가지고 깔끔하게 만들수도 있을거 같아요!
https://stackoverflow.com/questions/37972859/java-util-objects-isnull-vs-object-null

File: src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java (hunk 6 above)
Comment (ksy90101) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java:26]: 굳굳 👍

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java [hunk 8]
```diff
@@ -30,4 +35,30 @@ public List<LocalDate> convertFromScheduleListToDateList(List<Schedule> schedule
...
+		for (TheaterRoom theaterRoom : theaterRooms) {
+			List<LocalTime> startTimeListInTheaterRoom =
+				getStartTimeListOfSchedulesInTheaterRoom(schedules, theaterRoom);
+
+			responseFindTimeSchedules.add(new ResponseFindTime(theaterRoom.getName(),
+				theaterRoom.getTotalAmount(), startTimeListInTheaterRoom));
+		}
+
+		return responseFindTimeSchedules;
+	}
```
Comment (ksy90101) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java:54]: 스트림으로 가능할거 같은데...

Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/service/utils/ScheduleConverter.java:54]: 제 지능이 부족해서 몇시간 고민해도 잘 안되더라고요..

**Pull request merged**